RecommenderSystems/deepfm/deepfm_train_eval.py

```diff
@@ -0,0 +1,662 @@
+            init_scale=1073741824, growth_factor=2.0, backoff_factor=0.5, growth_interval=2000,
+        )
+    
+    eval_graph = DeepFMValGraph(deepfm_module, args.amp)
+    train_graph = DeepFMTrainGraph(deepfm_module, loss, opt, grad_scaler, args.amp)
+
+    train_losses = []
+    batches_per_epoch = math.ceil(args.num_train_samples / args.batch_size)
+
+    best_metric = -np.inf
+    stopping_steps = 0
+
+    deepfm_module.train()
+    step, last_step, last_time = -1, 0, time.time()
+    epoch = 0
+    with make_criteo_dataloader(f"{args.data_dir}/train", args.batch_size) as loader:
+        for step in range(1, args.train_batches + 1):
+            labels, features = batch_to_global(*next(loader))
+            loss = train_graph(labels, features)
+            loss = loss.numpy()
+            train_losses.append(loss)
+            if step % args.loss_print_interval == 0:
+                if rank == 0:
+                    latency_ms = 1000 * (time.time() - last_time) / (step - last_step)
+                    last_step, last_time = step, time.time()
+                    strtime = time.strftime("%Y-%m-%d %H:%M:%S")
+                    print(
+                        f"Rank[{rank}], Step {step}, Loss {loss:0.4f}, "
+                        + f"Latency {latency_ms:0.3f} ms, {strtime}"
+                    )
+
+            if step % batches_per_epoch == 0:
+                epoch += 1
+                auc, logloss = eval(args, eval_graph, tag='val', cur_step=step, epoch=epoch)
+                if args.save_model_after_each_eval:
+                    save_model(f"step_{step}_val_auc_{auc:0.5f}")
+
+                monitor_value = get_metrics(logs={'auc': auc, 'logloss': logloss})
+
+                stop_training, best_metric, stopping_steps, save_best = early_stop(
+                    epoch, 
+                    monitor_value, 
+                    best_metric=best_metric, 
+                    stopping_steps=stopping_steps, 
+                    patience=args.patience, 
+                    min_delta=args.min_delta,
+                )
+                if save_best:
+                    print(f"Save best model: monitor(max): {best_metric:.6f}")
+                    save_model("best_checkpoint")
+                if stop_training:
+                    break
+
+                deepfm_module.train()
+                last_time = time.time()
+
+    state_dict = flow.load(f"{args.model_save_dir}/best_checkpoint", global_src_rank=0)
+    deepfm_module.load_state_dict(state_dict, strict=False)
+    print("================ Validation Evaluation ================")
+    auc, logloss = eval(args, eval_graph, tag='val', cur_step=step, epoch=epoch)
+    print("================ Test Evaluation ================")
+    auc, logloss = eval(args, eval_graph, tag='test', cur_step=step, epoch=epoch)
+
+
+def batch_to_global(np_label, np_features):
+    def _np_to_global(np, dtype=flow.float):
+        t = flow.tensor(np, dtype=dtype)
```

Comment:
use flow.from_numpy can reduce `cast` time, but the dtype in dataset should match with target dtype.